File: Metodos Numericos/metodos_numericos.cpp
#include <iostream>
#include <fstream>
#include <cmath>
#include <stdexcept>
#define MAX_ITER 100

using namespace std;

double func(double x) {
     return pow(x, 3) - 9 * x + 3;
}

double funclinha(double x) {
    return 3 * pow(x, 2) - 9;
}

double phi(double x) {
    return ((pow(x, 3)) / 9) + (1.0 / 3.0);
}

void bisseccao_Metodo(double a, double b, double precisao, ofstream &arquivo) {
    arquivo << "---------| MÉTODO DA BISSEÇÃO |---------" << endl;
    int k = 0;
    double meio, fmeio, fa = func(a);

    if (fa * func(b) > 0) {
        arquivo << "Erro: f(a) e f(b) têm o mesmo sinal. Método não aplicável no intervalo [" 
                << a << ", " << b << "]." << endl << endl;
        return;
    }

    while ((fabs(b - a) > precisao) && (k < MAX_ITER)) {
        k++;
        meio = (a + b) / 2;
        fmeio = func(meio);

        arquivo << "Iteração " << k << ": x = " << meio << ", f(x) = " << fmeio << ", intervalo = [" 
                << a << ", " << b << "]" << endl;

        if (fmeio == 0) {
            arquivo << "Raiz exata encontrada: " << meio << endl << endl;
            return;
        }

        if (fa * fmeio < 0)
            b = meio;
        else {
            a = meio;
            fa = fmeio;
        }
    }
    arquivo << "Raiz aproximada: " << (a + b) / 2 << " após " << k << " iterações" << endl << endl;
}

void newton_Metodo(double x0, double precisao, ofstream &arquivo) {
    arquivo << "---------| MÉTODO DE NEWTON |---------" << endl;
    int k = 0;
    double x1, fx, fxlinha;

    do {
        fx = func(x0);
        fxlinha = funclinha(x0);
        
        if (fabs(fxlinha) < 1e-12) {
            arquivo << "Erro: Derivada próxima de zero (" << fxlinha << "). Método falhou na iteração " 
                   << k << " com x = " << x0 << endl << endl;
            return;
        }

        x1 = x0 - fx / fxlinha;
        k++;

        arquivo << "Iteração " << k << ": x = " << x1 << ", f(x) = " << func(x1) 
                << ", |xₙ - xₙ₋₁| = " << fabs(x1 - x0) << endl;

        if (fabs(x1 - x0) < precisao || fabs(func(x1)) < precisao)
            break;

        x0 = x1;
    } while (k < MAX_ITER);

    arquivo << "Raiz aproximada: " << x1 << " após " << k << " iterações" << endl << endl;
}

void mil_Metodo(double x0, double precisao, ofstream &arquivo) {
    arquivo << "---------| MÉTODO ITERATIVO LINEAR (MIL) |---------" << endl;
    int k = 0;
    double x1;
    bool convergencia_garantida = fabs(3 * pow(x0, 2) / 9) < 1; // |φ'(x)| < 1?

    if (!convergencia_garantida) {
        arquivo << "Atenção: Convergência não garantida para x0 = " << x0 
                << " (|φ'(x0)| = " << fabs(3 * pow(x0, 2) / 9) << ")" << endl;
    }

    do {
        x1 = phi(x0);
        k++;

        arquivo << "Iteração " << k << ": x = " << x1 << ", f(x) = " << func(x1) 
                << ", |xₙ - xₙ₋₁| = " << fabs(x1 - x0) << endl;

        if (fabs(x1 - x0) < precisao)
            break;

        x0 = x1;
    } while (k < MAX_ITER);

    arquivo << "Raiz aproximada: " << x1 << " após " << k << " iterações" << endl << endl;
}

void secante_Metodo(double x0, double x1, double precisao, ofstream &arquivo) {
    arquivo << "---------| MÉTODO DA SECANTE |---------" << endl;
    int k = 0;
    double x2, f0, f1;

    if (fabs(x1 - x0) < 1e-12) {
        arquivo << "Erro: x0 e x1 muito próximos. Método não pode ser aplicado." << endl << endl;
        return;
    }

    f0 = func(x0);
    f1 = func(x1);

    do {
        if (fabs(f1 - f0) < 1e-12) {
            arquivo << "Erro: Divisão por zero na iteração " << k 
                   << ". Diferença entre f(x1) e f(x0) muito pequena." << endl << endl;
            return;
        }

        x2 = x1 - (f1 * (x1 - x0)) / (f1 - f0);
        k++;

        arquivo << "Iteração " << k << ": x = " << x2 << ", f(x) = " << func(x2) 
                << ", |xₙ - xₙ₋₁| = " << fabs(x2 - x1) << endl;

        if (fabs(x2 - x1) < precisao)
            break;

        x0 = x1; f0 = f1;
        x1 = x2; f1 = func(x1);
    } while (k < MAX_ITER);

    arquivo << "Raiz aproximada: " << x2 << " após " << k << " iterações" << endl << endl;
}

void regulaFalsi_Metodo(double a, double b, double precisao, ofstream &arquivo) {
    arquivo << "---------| MÉTODO DA REGULA FALSI |---------" << endl;
    double fa = func(a), fb = func(b), x, fx;
    int k = 0;

    if (fa * fb > 0) {
        arquivo << "Erro: f(a) e f(b) têm o mesmo sinal. Método não aplicável no intervalo [" 
                << a << ", " << b << "]." << endl << endl;
        return;
    }

    do {
        x = (a * fb - b * fa) / (fb - fa);
        fx = func(x);
        k++;

        arquivo << "Iteração " << k << ": x = " << x << ", f(x) = " << fx 
                << ", intervalo = [" << a << ", " << b << "]" << endl;

        if (fabs(fx) < precisao)
            break;

        if (fa * fx < 0) {
            b = x;
            fb = fx;
        } else {
            a = x;
            fa = fx;
        }
    } while (k < MAX_ITER);

    arquivo << "Raiz aproximada: " << x << " após " << k << " iterações" << endl << endl;
}

// ===================== Função Principal =====================
int main() {
    ifstream entrada("entrada.txt");
    ofstream arquivo("saida.txt");

    if (!entrada.is_open()) {
        cerr << "Erro: Não foi possível abrir o arquivo de entrada 'entrada.txt'" << endl;
        return 1;
    }

    if (!arquivo.is_open()) {
        cerr << "Erro: Não foi possível abrir o arquivo de saída 'saida.txt'" << endl;
        return 1;
    }

    double a, b, x0, x1, precisao;
    entrada >> a >> b >> x0 >> x1 >> precisao;

    if (entrada.fail()) {
        cerr << "Erro: Formato inválido no arquivo de entrada. Esperado: a b x0 x1 precisao" << endl;
        return 1;
    }

    arquivo << "PARÂMETROS DE ENTRADA:" << endl;
    arquivo << "Intervalo inicial [a, b] = [" << a << ", " << b << "]" << endl;
    arquivo << "Chute inicial x0 = " << x0 << endl;
    arquivo << "Segundo ponto x1 = " << x1 << endl;
    arquivo << "Precisão requerida = " << precisao << endl << endl;

    bisseccao_Metodo(a, b, precisao, arquivo);
    mil_Metodo(x0, precisao, arquivo);
    newton_Metodo(x0, precisao, arquivo);
    secante_Metodo(x0, x1, precisao, arquivo);
    regulaFalsi_Metodo(a, b, precisao, arquivo);

    cout << "Cálculos concluídos. Resultados gravados em 'saida.txt'" << endl;

    entrada.close();
    arquivo.close();
    return 0;
}

File: Metodos Numericos/metodos_numericos.py
import sympy as sp

MAX_ITER = 100

# Formato do txt:
# função
# função_phi
# a b x0 x1 precisao

#Variáveis globais
x = sp.symbols('x')
func_expr = None
deriv_expr = None
phi_expr = None
precisao = None
arquivo = None

def func(val):
    return float(func_expr.subs(x, val))

def funclinha(val):
    return float(deriv_expr.subs(x, val))

def phi(val):
    return float(phi_expr.subs(x, val))

def condicaoParada(x0, x1):
    if abs(x1 - x0) <= precisao or abs(func(x1)) <= precisao:
        arquivo.write("\nCriterio de parada atingido (Precisao = {}):\n".format(precisao))
        arquivo.write("|{} - {}| = {}\n".format(x1, x0, abs(x1 - x0)))
        arquivo.write("|f({})| = {}\n".format(x1, abs(func(x1))))
        arquivo.write("\n")
        return True
    return False

# MÉTODO DA BISSECÇÃO (isso nao foi gerado por GPT)
# 1. Calcula o ponto médio entre de [a, b]
# 2. Verifica-se os sinais de f(m)
# 3. Dependendo do sinal de f(m), substitui-se o antigo intervalo pelo novo
# 4. Repita o processo até atingir o critério de parada
def bisseccao(a, b):
    arquivo.write("---------| METODO DA BISSECAO |---------\n")
    k = 0
    fa, fb = func(a), func(b)
    meio = None
    # Se não existir a troca de sinal, não existe um ponto entre a e b que é zero (raiz)
    if fa * fb > 0:
        arquivo.write("Erro: f(a) e f(b) tem o mesmo sinal.\n\n")
        return
    while not condicaoParada(a, b) and k < MAX_ITER:
        k += 1
        meio = 0.5 * (a + b)
        fmeio = func(meio)
        arquivo.write(f"Iteracao {k}: x = {meio}, f(x) = {fmeio}, intervalo = [{a}, {b}]\n")
        if fa * fmeio < 0:
            b, fb = meio, fmeio
        else:
            a, fa = meio, fmeio
    arquivo.write(f"Raiz aproximada: {meio} apos {k} iteracoes.\n\n")

# MÉTODO ITERATIVO LINEAR (agora nao pode comentar nos codigo pq é tudo GPT hj em dia)
# 1. Lê a função Phi
# 2. Utiliza um valor inicial x0
# 3. Calcula xk+1 = phi(xk)
# 4. Repita o processo até atingir o critério de parada
def mil(x0):
    arquivo.write("---------| METODO ITERATIVO LINEAR (MIL) |---------\n")
    k = 0
    if abs(func(x0)) < precisao:
        return
    while k < MAX_ITER:
        x1 = phi(x0)
        k += 1
        arquivo.write(f"Iteracao {k}: x = {x1}\n")
        if condicaoParada(x0, x1):
            break
        x0 = x1
    arquivo.write(f"Raiz aproximada: {x1} apos {k} iteracoes\n\n")

# MÉTODO DE NEWTON
# 1. Escolhe um valor inicial x0
# 2. Calcula f(x0) e f'(x0)
# 3. Calcula x1 = x0 - f(x0)/f'(x0)
# 4. Repita o processo até atingir o critério de parada
def newton(x0):
    arquivo.write("---------| METODO DE NEWTON |---------\n")
    k = 0
    while k < MAX_ITER:
        fx, fxlinha = func(x0), funclinha(x0)
        if fxlinha == 0:
            arquivo.write(f"Erro: derivada zero em {x0}\n")
            return
        x1 = x0 - fx/fxlinha
        k += 1
        arquivo.write(f"Iteracao {k}: x = {x1}\n")
        if condicaoParada(x0, x1):
            break
        x0 = x1
    arquivo.write(f"Raiz aproximada: {x1} apos {k} iteracoes\n\n")

# MÉTODO DA SECANTE
# 1. Escolhe dois valores iniciais x0 e x1
# 2. Calcula f(x0) e f(x1)
# 3. Calcula x2 = x1 - f(x1)*(x1 - x0)/(f(x1) - f(x0))
# 4. Repita o processo até atingir o critério de parada
def secante(x0, x1):
    arquivo.write("---------| METODO DA SECANTE |---------\n")
    k = 0
    f0, f1 = func(x0), func(x1)
    while k < MAX_ITER:
        if f1 - f0 == 0:
            arquivo.write(f"Erro: divisao por zero\n")
            return
        x2 = x1 - f1 * (x1 - x0) / (f1 - f0)
        k += 1
        arquivo.write(f"Iteracao {k}: x = {x2}\n")
        if condicaoParada(x1, x2):
            break
        x0, f0 = x1, f1
        x1, f1 = x2, func(x2)
    arquivo.write(f"Raiz aproximada: {x2} apos {k} iteracoes\n\n")

# MÉTODO DA REGULA FALSI
# 1. Escolhe um intervalo [a, b]
# 2. Calcula o ponto de interseção x = (a*fb - b*fa)/(fb - fa)
# 3. Verifica o critério de parada (|x - a| ou |f(x)|)
# 4. Atualiza o intervalo:
#    - se f(a)*f(x) < 0, então b = x
#    - senão, a = x
# 5. Repete o processo até convergir
def regulaFalsi(a, b):
    arquivo.write("---------| METODO DA REGULA FALSI |---------\n")
    fa, fb = func(a), func(b)
    if fa * fb > 0:
        arquivo.write("Erro: f(a) e f(b) tem o mesmo sinal\n\n")
        return
    k = 0
    while k < MAX_ITER:
        x = (a * fb - b * fa) / (fb - fa)
        fx = func(x)
        k += 1
        arquivo.write(f"Iteracao {k}: x = {x}, f(x) = {fx}, intervalo = [{a}, {b}]\n")
        if condicaoParada(a, x):
            break
        if fa * fx < 0:
            b, fb = x, fx
        else:
            a, fa = x, fx
    arquivo.write(f"Raiz aproximada: {x} apos {k} iteracoes\n\n")

def main():
    global func_expr, deriv_expr, phi_expr, precisao, arquivo
    try:
        with open("entrada.txt", "r") as f:
            lines = f.read().splitlines()
            locals_dict = {"log": lambda arg: sp.log(arg, 10)}

            # Função
            func_expr = sp.sympify(lines[0], locals=locals_dict)

            # Phi
            phi_expr = sp.sympify(lines[1], locals=locals_dict)

            #Calcula derivada
            deriv_expr = sp.diff(func_expr, x)

            #Lê os parâmetros 
            a, b, x0, x1, precisao = map(float, lines[2].split())

    except Exception as e:
        print("Erro ao ler entrada.txt:", e)
        return

    arquivo = open("saida.txt", "w")
    arquivo.write("FUNCAO: f(x) = {}\n".format(func_expr))
    arquivo.write("DERIVADA: f'(x) = {}\n".format(deriv_expr))
    arquivo.write("PHI(x): {}\n\n".format(phi_expr))
    arquivo.write(f"Intervalo [a,b] = [{a},{b}]\n")
    arquivo.write(f"x0 = {x0}, x1 = {x1}, precisao = {precisao}\n\n")

    bisseccao(a, b)
    mil(x0)
    newton(x0)
    secante(x0, x1)
    regulaFalsi(a, b)

    arquivo.close()
    print("Cálculo concluído. Veja saida.txt")

if __name__ == "__main__":
    main()